service/history/workflow/workflow_test/mutable_state_impl_test.go

```diff
@@ -0,0 +1,363 @@
+// but test packages force you to only test exported methods.
+// See https://github.com/maratori/testpackage#motivation for more on the rationale used here.
+package workflow_test
+
+import (
+	"fmt"
+	"math"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/golang/mock/gomock"
+	commonpb "go.temporal.io/api/common/v1"
+	enumspb "go.temporal.io/api/enums/v1"
+	"go.temporal.io/api/history/v1"
+	taskqueuepb "go.temporal.io/api/taskqueue/v1"
+	"go.temporal.io/api/workflowservice/v1"
+	"go.temporal.io/server/api/enums/v1"
+	"go.temporal.io/server/api/historyservice/v1"
+	"go.temporal.io/server/common/clock"
+	"go.temporal.io/server/common/cluster"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/common/namespace"
+	"go.temporal.io/server/common/persistence"
+	"go.temporal.io/server/common/primitives/timestamp"
+	"go.temporal.io/server/service/history/configs"
+	"go.temporal.io/server/service/history/events"
+	"go.temporal.io/server/service/history/shard"
+	"go.temporal.io/server/service/history/tests"
+	"go.temporal.io/server/service/history/workflow"
+)
+
+func TestMutableStateImpl_CloseTransactionAsMutation(t *testing.T) {
+	t.Parallel()
+
+	for _, tc := range []mutationTestCase{
+		{
+			name:              "numEvents<maxNumEvents",
+			numEvents:         1,
+			maxNumEvents:      2,
+			transactionPolicy: workflow.TransactionPolicyActive,
+			expectFailure:     false,
+		},
+		{
+			name:              "numEvents=maxNumEvents",
+			numEvents:         2,
+			maxNumEvents:      2,
+			transactionPolicy: workflow.TransactionPolicyActive,
+			expectFailure:     true,
+		},
+		{
+			name:              "numEvents>maxNumEvents",
+			numEvents:         3,
+			maxNumEvents:      2,
+			transactionPolicy: workflow.TransactionPolicyActive,
+			expectFailure:     true,
+		},
+	} {
+		t.Run(tc.name, tc.Run)
+	}
+}
+
+type mutationTestCase struct {
+	name              string
+	transactionPolicy workflow.TransactionPolicy
+	numEvents         int
+	maxNumEvents      int
+	expectFailure     bool
+}
+
+func (c mutationTestCase) Run(t *testing.T) {
+	t.Parallel()
+
+	nsEntry := tests.LocalNamespaceEntry
+	ms := c.createMutableState(t, nsEntry)
+
+	_, err := ms.AddWorkflowExecutionStartedEvent(
+		commonpb.WorkflowExecution{
+			WorkflowId: "694B31C3-1FDC-4C9E-87BC-747D539BF0CD",
+			RunId:      "3E1836B8-8692-440D-9495-45D9EABDED6B",
+		},
+		&historyservice.StartWorkflowExecutionRequest{
+			Attempt:     1,
+			NamespaceId: nsEntry.ID().String(),
+			StartRequest: &workflowservice.StartWorkflowExecutionRequest{
+				WorkflowType:        &commonpb.WorkflowType{Name: "workflow-type"},
+				TaskQueue:           &taskqueuepb.TaskQueue{Name: "task-queue-name"},
+				WorkflowRunTimeout:  timestamp.DurationPtr(200 * time.Second),
+				WorkflowTaskTimeout: timestamp.DurationPtr(1 * time.Second),
+			},
+		},
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	c.completeOneWFT(t, ms)
```

Comment:
why do we need to complete one wft first?

Reply:
I think I just need to close one mutation transaction first? I'm not sure why, but I simplified this to just this:

```go
	_, err := ms.AddWorkflowExecutionStartedEvent(
		commonpb.WorkflowExecution{
			WorkflowId: "694B31C3-1FDC-4C9E-87BC-747D539BF0CD",
			RunId:      "3E1836B8-8692-440D-9495-45D9EABDED6B",
		},
		&historyservice.StartWorkflowExecutionRequest{
			Attempt:     1,
			NamespaceId: nsEntry.ID().String(),
			StartRequest: &workflowservice.StartWorkflowExecutionRequest{
				WorkflowType:        &commonpb.WorkflowType{Name: "workflow-type"},
				TaskQueue:           &taskqueuepb.TaskQueue{Name: "task-queue-name"},
				WorkflowRunTimeout:  timestamp.DurationPtr(200 * time.Second),
				WorkflowTaskTimeout: timestamp.DurationPtr(1 * time.Second),
			},
		},
	)
	if err != nil {
		t.Fatal(err)
	}

	_, _, err = ms.CloseTransactionAsMutation(c.transactionPolicy)
	if err != nil {
		t.Fatal(err)
	}
```